File: cmd/roost/theme.go
package main

import (
	"bufio"
	"embed"
	"fmt"
	"io"
	"strconv"
	"strings"

	"github.com/charliek/roost/internal/ghostty"
)

// bundledThemes ships the user-selectable color schemes. Filenames mirror
// ghostty's themes/ directory exactly (no extension, spaces and `+`
// preserved) so users can copy theme files in from
// /Applications/Ghostty.app/Contents/Resources/ghostty/themes/ without
// renaming.
//
//go:embed all:themes
var bundledThemes embed.FS

// Theme is the set of colors Roost installs on every new terminal. The
// shape mirrors a ghostty theme file: a 256-entry palette plus the
// renderer-side colors libghostty doesn't model (selection, cursor-text,
// bold-color).
type Theme struct {
	Foreground          ghostty.ColorRGB
	Background          ghostty.ColorRGB
	Cursor              ghostty.ColorRGB
	CursorText          ghostty.ColorRGB
	BoldColor           ghostty.ColorRGB
	SelectionBackground ghostty.ColorRGB
	SelectionForeground ghostty.ColorRGB
	Palette             [256]ghostty.ColorRGB
}

// DefaultTheme is the bundled "roost-dark" theme parsed at init. It is
// the runtime fallback when the user names an unknown theme. A parse
// failure here is a build error: the embedded file is broken.
var DefaultTheme = mustLoadTheme("roost-dark")

// LoadTheme parses a bundled ghostty-format theme file by name. Names
// match the embedded filenames (e.g. "Dracula+", "Catppuccin Mocha").
// Returns an error if the theme isn't bundled or fails to parse.
func LoadTheme(name string) (Theme, error) {
	f, err := bundledThemes.Open("themes/" + name)
	if err != nil {
		return Theme{}, fmt.Errorf("theme %q: %w", name, err)
	}
	defer f.Close()
	th, err := parseTheme(f)
	if err != nil {
		return Theme{}, fmt.Errorf("theme %q: %w", name, err)
	}
	return th, nil
}

// BundledThemeNames returns the list of embedded theme names, sorted.
// Used by docs / debug output.
func BundledThemeNames() []string {
	entries, err := bundledThemes.ReadDir("themes")
	if err != nil {
		return nil
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		names = append(names, e.Name())
	}
	return names
}

func mustLoadTheme(name string) Theme {
	th, err := LoadTheme(name)
	if err != nil {
		panic(fmt.Errorf("bundled %s", err))
	}
	return th
}

// parseTheme reads a ghostty-format theme file. Format: `key = value`
// with `#` comments, identical to roost's main config syntax. Unknown
// keys are ignored (matches ghostty's tolerant loader); bad values
// produce errors. Indices 16-255 of the palette are not user-controlled
// in theme files — fillPalette240 writes the standard 6×6×6 cube and
// 24-step gray ramp after parsing.
func parseTheme(r io.Reader) (Theme, error) {
	var th Theme
	var sawBackground, sawForeground, sawCursor bool
	var sawCursorText, sawBoldColor bool
	var sawSelectionBackground, sawSelectionForeground bool

	sc := bufio.NewScanner(r)
	lineNum := 0
	for sc.Scan() {
		lineNum++
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, val, ok := splitThemeKV(line)
		if !ok {
			return Theme{}, fmt.Errorf("line %d: expected key = value", lineNum)
		}
		switch key {
		case "background":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: background: %w", lineNum, err)
			}
			th.Background = c
			sawBackground = true
		case "foreground":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: foreground: %w", lineNum, err)
			}
			th.Foreground = c
			sawForeground = true
		case "cursor-color":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: cursor-color: %w", lineNum, err)
			}
			th.Cursor = c
			sawCursor = true
		case "cursor-text":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: cursor-text: %w", lineNum, err)
			}
			th.CursorText = c
			sawCursorText = true
		case "bold-color":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: bold-color: %w", lineNum, err)
			}
			th.BoldColor = c
			sawBoldColor = true
		case "selection-background":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: selection-background: %w", lineNum, err)
			}
			th.SelectionBackground = c
			sawSelectionBackground = true
		case "selection-foreground":
			c, err := parseHexColor(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: selection-foreground: %w", lineNum, err)
			}
			th.SelectionForeground = c
			sawSelectionForeground = true
		case "palette":
			idx, c, err := parsePaletteEntry(val)
			if err != nil {
				return Theme{}, fmt.Errorf("line %d: palette: %w", lineNum, err)
			}
			// 16-255 are computed; ignore any explicit overrides so we
			// don't end up with a half-customized palette.
			if idx < 16 {
				th.Palette[idx] = c
			}
		}
		// Unknown keys are silently ignored.
	}
	if err := sc.Err(); err != nil {
		return Theme{}, fmt.Errorf("read: %w", err)
	}

	// Required fields. Without these the theme would render unreadable
	// (zero-value black on black). Bundled themes always set both;
	// failing fast catches a broken theme file at LoadTheme time.
	if !sawBackground {
		return Theme{}, fmt.Errorf("missing required key: background")
	}
	if !sawForeground {
		return Theme{}, fmt.Errorf("missing required key: foreground")
	}

	// Optional fields — fall back to documented defaults so every
	// Theme value is renderable without a nil check. Presence is
	// tracked by bool, not by zero-value comparison: an explicit
	// "selection-background = #000000" must round-trip as black, not
	// be treated as unset.
	if !sawCursor {
		th.Cursor = th.Foreground
	}
	if !sawCursorText {
		th.CursorText = th.Background
	}
	if !sawBoldColor {
		th.BoldColor = th.Foreground
	}
	if !sawSelectionBackground {
		th.SelectionBackground = th.Foreground
	}
	if !sawSelectionForeground {
		th.SelectionForeground = th.Background
	}

	fillPalette240(&th.Palette)
	return th, nil
}

// fillPalette240 writes indices 16-255: the standard xterm 6×6×6 RGB
// cube (16-231) and 24-step gray ramp (232-255). Algorithm matches
// ../ghostty/src/terminal/color.zig:7-45.
func fillPalette240(p *[256]ghostty.ColorRGB) {
	i := 16
	for r := 0; r < 6; r++ {
		for g := 0; g < 6; g++ {
			for b := 0; b < 6; b++ {
				p[i] = ghostty.ColorRGB{
					R: cubeAxis(r),
					G: cubeAxis(g),
					B: cubeAxis(b),
				}
				i++
			}
		}
	}
	for n := 0; n < 24; n++ {
		v := uint8(n*10 + 8)
		p[232+n] = ghostty.ColorRGB{R: v, G: v, B: v}
	}
}

func cubeAxis(n int) uint8 {
	if n == 0 {
		return 0
	}
	return uint8(n*40 + 55)
}

// parseHexColor accepts "#RRGGBB" (case-insensitive).
func parseHexColor(s string) (ghostty.ColorRGB, error) {
	s = strings.TrimSpace(s)
	if len(s) != 7 || s[0] != '#' {
		return ghostty.ColorRGB{}, fmt.Errorf("expected #RRGGBB, got %q", s)
	}
	r, err := strconv.ParseUint(s[1:3], 16, 8)
	if err != nil {
		return ghostty.ColorRGB{}, fmt.Errorf("bad red in %q: %w", s, err)
	}
	g, err := strconv.ParseUint(s[3:5], 16, 8)
	if err != nil {
		return ghostty.ColorRGB{}, fmt.Errorf("bad green in %q: %w", s, err)
	}
	b, err := strconv.ParseUint(s[5:7], 16, 8)
	if err != nil {
		return ghostty.ColorRGB{}, fmt.Errorf("bad blue in %q: %w", s, err)
	}
	return ghostty.ColorRGB{R: uint8(r), G: uint8(g), B: uint8(b)}, nil
}

// parsePaletteEntry parses a `palette` value: "N=#RRGGBB" where N is
// 0-255. Returns the index and color.
func parsePaletteEntry(val string) (int, ghostty.ColorRGB, error) {
	eq := strings.IndexByte(val, '=')
	if eq < 0 {
		return 0, ghostty.ColorRGB{}, fmt.Errorf("expected N=#RRGGBB, got %q", val)
	}
	idx, err := strconv.Atoi(strings.TrimSpace(val[:eq]))
	if err != nil {
		return 0, ghostty.ColorRGB{}, fmt.Errorf("index in %q: %w", val, err)
	}
	if idx < 0 || idx > 255 {
		return 0, ghostty.ColorRGB{}, fmt.Errorf("index out of range [0,255]: %d", idx)
	}
	c, err := parseHexColor(strings.TrimSpace(val[eq+1:]))
	if err != nil {
		return 0, ghostty.ColorRGB{}, err
	}
	return idx, c, nil
}

// splitThemeKV mirrors config.splitKV: trim whitespace around `=`,
// strip an optional pair of surrounding double quotes from the value.
func splitThemeKV(line string) (key, val string, ok bool) {
	eq := strings.IndexByte(line, '=')
	if eq < 0 {
		return "", "", false
	}
	key = strings.TrimSpace(line[:eq])
	val = strings.TrimSpace(line[eq+1:])
	if len(val) >= 2 && val[0] == '"' && val[len(val)-1] == '"' {
		val = val[1 : len(val)-1]
	}
	return key, val, key != ""
}

// DefaultDeviceAttrs is what libghostty advertises when programs probe
// the terminal type via CSI c / CSI > c / CSI = c. Conservative
// xterm-256color-ish values: VT220 conformance, ANSI color, no sixel.
// Programs that gate on these will enable the modern feature set.
var DefaultDeviceAttrs = ghostty.DeviceAttrs{
	// DA1: \e[?62;22c — VT220 + ANSI_COLOR.
	PrimaryConformance: 62, // VT220
	PrimaryFeatures:    []uint16{22 /* ANSI_COLOR */},

	// DA2: \e[>0;276;0c — VT100 device, version 276 (matches xterm.js
	// parity placeholder used by similar terminals), no ROM cartridge.
	SecondaryDeviceType:      0,
	SecondaryFirmwareVersion: 276,
	SecondaryRomCartridge:    0,

	// DA3: programs that ask for this mostly check it exists; the value
	// is rendered as 8 hex digits (DECRPTUI). Zero is fine.
	TertiaryUnitID: 0,
}

File: mac/Sources/Roost/Theme.swift
// Roost theme — Phase 6a M6.
//
// The theme system — same theme files (same names, same on-disk format)
// as the Rust/Linux UI. The bundled theme files live at
// `Sources/Roost/Resources/themes/`; the source-of-truth copy is in the
// Rust crate at `crates/roost-linux/src/resources/themes/`, and the two
// trees are kept byte-identical by `make themes-check`. Add a new theme
// to both trees (and to `BUNDLED_THEMES` in `theme.rs`).
//
// File format mirrors Ghostty's `themes/` directory entries
// (e.g. `palette = 0=#1a1a1a`, `background = #1e1e1e`,
// `selection-background = #3f638b`, …). We deliberately don't parse
// every key Ghostty's renderer recognizes — only the ones Roost
// actually applies in `TerminalView.draw(_:)` and the window chrome.

import AppKit
import CGhosttyVT
import Foundation

extension Bundle {
    /// The bundle carrying Roost's themes (`themes/`). In a packaged
    /// `.app` it lives at `Contents/Resources/Roost_Roost.bundle`,
    /// resolved here via `Bundle.main` — deliberately NOT via SwiftPM's
    /// generated `Bundle.module`. That accessor searches exactly
    /// `Bundle.main.bundleURL/Roost_Roost.bundle` (the `.app` ROOT) plus
    /// a build-machine path baked in at compile time; the root can't be
    /// populated (nested bundles outside `Contents/` break codesigning)
    /// and the build path doesn't exist on a user's machine, so
    /// `Bundle.module` `fatalError`ed on every clean install while
    /// dev/CI launches — which have that build path present — passed.
    /// Falls back to `Bundle.module` for `swift run` / tests, where no
    /// `.app` exists and the build-tree bundle does resolve.
    static var roostResources: Bundle {
        // `resourceURL` is `Contents/Resources` for a normally-launched
        // `.app`; the explicit `bundleURL/Contents/Resources` candidate
        // covers direct-exec contexts (e.g. running the binary by path),
        // where `resourceURL` can resolve elsewhere. First hit wins.
        for candidate in [
            Bundle.main.resourceURL?.appendingPathComponent("Roost_Roost.bundle"),
            Bundle.main.bundleURL
                .appendingPathComponent("Contents/Resources/Roost_Roost.bundle"),
        ] {
            if let url = candidate, let bundle = Bundle(url: url) { return bundle }
        }
        return .module
    }
}

/// Parsed terminal color scheme. Mirrors the Go binary's `Theme`
/// struct field-for-field so a future "share theme between Go and
/// Swift binaries" diagnostic stays trivial.
struct Theme: Sendable {
    var foreground: NSColor
    var background: NSColor
    var cursor: NSColor
    var selectionBackground: NSColor
    var selectionForeground: NSColor
    /// Ghostty `bold-color` accent. `nil` until a theme opts in;
    /// bold cells without an explicit SGR fg adopt this color when
    /// set, matching the Linux `Theme.bold_color` shape. Defaulted
    /// here so existing memberwise-init callers (smoke tests, future
    /// fixtures) keep compiling without naming the field.
    var boldColor: NSColor? = nil
    var palette: [NSColor]  // exactly 256 entries

    /// Hard-coded fallback so the UI has a sane theme before
    /// `loadBundled(name:)` runs (or when it errors). Mirrors the
    /// "roost-dark" theme's headline colors.
    static let fallback: Theme = .init(
        foreground: NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 1),
        background: NSColor(srgbRed: 0.118, green: 0.118, blue: 0.118, alpha: 1),
        cursor: NSColor(srgbRed: 0.6, green: 0.6, blue: 0.6, alpha: 1),
        selectionBackground: NSColor(srgbRed: 0.247, green: 0.388, blue: 0.545, alpha: 1),
        selectionForeground: NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 1),
        boldColor: nil,
        palette: Theme.standardXterm256Palette()
    )

    /// The standard xterm 256-color palette: 16 ANSI colors (0–15), the
    /// 6×6×6 color cube (16–231), and the 24-step grayscale ramp
    /// (232–255). Used as the palette BASE that theme files override on
    /// top of — a theme file only specifies the 16 ANSI slots, so without
    /// this base every 256-color index (`SGR 48;5;N` / `38;5;N`) fell back
    /// to a flat placeholder and rendered the same wrong color. 256-color
    /// TUIs (vim/htop/lazygit, and opencode over SSH where COLORTERM is
    /// unset) depend on the cube + ramp being correct. Matches libghostty's
    /// and xterm's computed values.
    static func standardXterm256Palette() -> [NSColor] {
        func srgb(_ r: Int, _ g: Int, _ b: Int) -> NSColor {
            NSColor(srgbRed: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: 1)
        }
        var p: [NSColor] = []
        p.reserveCapacity(256)
        // 0–15: standard ANSI (normal + bright).
        let ansi = [
            (0, 0, 0), (128, 0, 0), (0, 128, 0), (128, 128, 0),
            (0, 0, 128), (128, 0, 128), (0, 128, 128), (192, 192, 192),
            (128, 128, 128), (255, 0, 0), (0, 255, 0), (255, 255, 0),
            (0, 0, 255), (255, 0, 255), (0, 255, 255), (255, 255, 255),
        ]
        for (r, g, b) in ansi { p.append(srgb(r, g, b)) }
        // 16–231: 6×6×6 color cube. Channel levels are 0, then 95+40·n.
        let levels = [0, 95, 135, 175, 215, 255]
        for i in 0..<216 {
            p.append(srgb(levels[(i / 36) % 6], levels[(i / 6) % 6], levels[i % 6]))
        }
        // 232–255: 24-step grayscale ramp, 8 + 10·n.
        for i in 0..<24 {
            let v = 8 + i * 10
            p.append(srgb(v, v, v))
        }
        return p
    }

    /// Look up a theme by bundled file name. Names match the
    /// embedded filenames verbatim (`"Dracula+"`, `"Catppuccin
    /// Mocha"`, `"roost-dark"`, …); no extension. Returns the
    /// fallback theme on any parse / resource-lookup error so the
    /// UI never goes color-less.
    @MainActor
    static func loadBundled(name: String) -> Theme {
        guard let url = Bundle.roostResources.url(
            forResource: name,
            withExtension: nil,
            subdirectory: "themes"
        ) else {
            NSLog("roost-mac: theme %@ not found in bundle; using fallback", name)
            return .fallback
        }
        do {
            let data = try String(contentsOf: url, encoding: .utf8)
            return try parse(data)
        } catch {
            NSLog("roost-mac: theme %@ parse failed: %@", name, "\(error)")
            return .fallback
        }
    }

    /// Returns the list of bundled theme names (sorted). Useful for
    /// future "themes" picker UI; M6 doesn't use it but the
    /// equivalent `BundledThemeNames()` lives on the Go side and
    /// keeps the surface parallel.
    @MainActor
    static func bundledNames() -> [String] {
        guard let urls = Bundle.roostResources.urls(
            forResourcesWithExtension: nil,
            subdirectory: "themes"
        ) else { return [] }
        return urls.map { $0.lastPathComponent }.sorted()
    }
}

/// Parse a Ghostty-format theme file. Same line-per-key shape as
/// `cmd/roost/theme.go::parseTheme`:
///   palette = N=#RRGGBB
///   background = #RRGGBB
///   foreground = #RRGGBB
///   cursor-color = #RRGGBB
///   selection-background = #RRGGBB
///   selection-foreground = #RRGGBB
/// Unknown keys are dropped (forward-compat with Ghostty additions).
/// Blank lines + `#` comments are skipped. Throws on syntactically
/// broken color literals.
private func parse(_ text: String) throws -> Theme {
    var theme = Theme.fallback
    var palette = theme.palette
    for raw in text.split(separator: "\n", omittingEmptySubsequences: false) {
        let line = raw.trimmingCharacters(in: .whitespaces)
        if line.isEmpty || line.hasPrefix("#") { continue }
        guard let eqIdx = line.firstIndex(of: "=") else { continue }
        let key = line[..<eqIdx].trimmingCharacters(in: .whitespaces)
        let rest = line[line.index(after: eqIdx)...]
            .trimmingCharacters(in: .whitespaces)
        switch key {
        case "palette":
            // `palette = N=#RRGGBB`
            guard let inner = rest.firstIndex(of: "=") else { continue }
            let nStr = rest[..<inner].trimmingCharacters(in: .whitespaces)
            let cStr = rest[rest.index(after: inner)...]
                .trimmingCharacters(in: .whitespaces)
            if let n = Int(nStr), n >= 0, n < 256,
               let c = parseHexColor(cStr)
            {
                palette[n] = c
            }
        case "background":
            if let c = parseHexColor(rest) { theme.background = c }
        case "foreground":
            if let c = parseHexColor(rest) { theme.foreground = c }
        case "cursor-color":
            if let c = parseHexColor(rest) { theme.cursor = c }
        case "selection-background":
            if let c = parseHexColor(rest) { theme.selectionBackground = c }
        case "selection-foreground":
            if let c = parseHexColor(rest) { theme.selectionForeground = c }
        case "bold-color":
            if let c = parseHexColor(rest) { theme.boldColor = c }
        default:
            // Ghostty has many more keys (cursor-text, link-color, …).
            // Drop them silently rather than erroring so a user can
            // keep an unchanged Ghostty theme file on disk and have
            // Roost honor what it can.
            continue
        }
    }
    theme.palette = palette
    return theme
}

// MARK: - libghostty-vt application (Phase 6a P3)

extension Theme {
    /// Colors pre-converted to libghostty's RGB structs. Converting the
    /// 256-entry palette runs `NSColor.usingColorSpace(.sRGB)` 256 times;
    /// the command palette's live theme preview broadcasts one theme to
    /// every open terminal on each arrow keypress, so we resolve once
    /// (`resolved()`) and reuse the result per terminal rather than
    /// re-converting N times.
    struct Resolved {
        var foreground: GhosttyColorRgb
        var background: GhosttyColorRgb
        var cursor: GhosttyColorRgb
        var palette: [GhosttyColorRgb]  // exactly 256 entries
    }

    @MainActor
    func resolved() -> Resolved {
        var palette = [GhosttyColorRgb](repeating: ghosttyColor(.black), count: 256)
        for i in 0..<min(self.palette.count, 256) {
            palette[i] = ghosttyColor(self.palette[i])
        }
        return Resolved(
            foreground: ghosttyColor(foreground),
            background: ghosttyColor(background),
            cursor: ghosttyColor(cursor),
            palette: palette
        )
    }

    /// Push fg / bg / cursor + 256-color palette into a libghostty-vt
    /// terminal. Mirrors the Go binary's
    /// `internal/ghostty/terminal.go::SetTheme`.
    ///
    /// Safe to call at any point in a terminal's life, including after
    /// `ghostty_terminal_vt_write` — the command palette's live theme
    /// preview relies on this, and `themeAppliesAfterVtWrite` pins it.
    /// The palette set preserves any per-index OSC overrides a program
    /// applied, so OSC-overridden / truecolor cells keep their colors.
    ///
    /// libghostty-vt's `GHOSTTY_TERMINAL_OPT_COLOR_*` options take a
    /// pointer to a `GhosttyColorRgb` (or `[256]GhosttyColorRgb` for the
    /// palette), read synchronously inside the call — the local storage
    /// doesn't need to outlive the call.
    @MainActor
    static func apply(_ colors: Resolved, to terminal: GhosttyTerminal) {
        func set(_ option: GhosttyTerminalOption, _ value: UnsafeMutableRawPointer, _ what: StaticString) {
            let rc = ghostty_terminal_set(terminal, option, value)
            if rc.rawValue != 0 {
                NSLog("roost-mac: ghostty_terminal_set(%@) failed rc=%d", "\(what)", rc.rawValue)
            }
        }
        var fg = colors.foreground
        set(GHOSTTY_TERMINAL_OPT_COLOR_FOREGROUND, &fg, "foreground")
        var bg = colors.background
        set(GHOSTTY_TERMINAL_OPT_COLOR_BACKGROUND, &bg, "background")
        var cursor = colors.cursor
        set(GHOSTTY_TERMINAL_OPT_COLOR_CURSOR, &cursor, "cursor")
        // The palette is a contiguous 256-entry C array; pass its base
        // pointer. The 256 count matches libghostty's signature exactly;
        // sending fewer is undefined behaviour per the header docs.
        var palette = colors.palette
        palette.withUnsafeMutableBufferPointer { buf in
            guard let base = buf.baseAddress else { return }
            set(GHOSTTY_TERMINAL_OPT_COLOR_PALETTE, UnsafeMutableRawPointer(base), "palette")
        }
    }

    /// Convenience for the single-terminal path (terminal init): resolve
    /// the theme's colors and apply them in one step.
    @MainActor
    static func apply(_ theme: Theme, to terminal: GhosttyTerminal) {
        apply(theme.resolved(), to: terminal)
    }
}

/// Convert an NSColor (in any color space) to libghostty's RGB struct.
/// Converts to sRGB first to match the bit-exact values the Go binary
/// pushes; `usingColorSpace(.sRGB)` returns nil for color spaces that
/// can't be represented in sRGB (rare for the named colors we use),
/// in which case we fall back to a black pixel rather than corrupting
/// the FFI struct with NaN-derived bytes.
@MainActor
private func ghosttyColor(_ color: NSColor) -> GhosttyColorRgb {
    let srgb = color.usingColorSpace(.sRGB) ?? .black
    let r = UInt8((srgb.redComponent * 255.0).rounded().clamped(to: 0...255))
    let g = UInt8((srgb.greenComponent * 255.0).rounded().clamped(to: 0...255))
    let b = UInt8((srgb.blueComponent * 255.0).rounded().clamped(to: 0...255))
    return GhosttyColorRgb(r: r, g: g, b: b)
}

private extension Comparable {
    /// Clamp `self` into a closed range. Generic helper because Swift
    /// stdlib's only built-in clamp lives on `Strideable` ranges, not
    /// the closed-range arithmetic the color conversion above wants.
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}

/// Parse `#RRGGBB` or `#RGB`. Lenient: returns nil for invalid
/// strings rather than throwing, so a stray field doesn't kill the
/// whole theme load.
private func parseHexColor(_ s: String) -> NSColor? {
    var hex = s
    if hex.hasPrefix("#") { hex.removeFirst() }
    guard let v = UInt32(hex, radix: 16) else { return nil }
    switch hex.count {
    case 6:
        let r = CGFloat((v >> 16) & 0xff) / 255.0
        let g = CGFloat((v >> 8) & 0xff) / 255.0
        let b = CGFloat(v & 0xff) / 255.0
        return NSColor(srgbRed: r, green: g, blue: b, alpha: 1)
    case 3:
        // RGB short form — each nibble expands to a pair.
        let r = CGFloat(((v >> 8) & 0xf) * 0x11) / 255.0
        let g = CGFloat(((v >> 4) & 0xf) * 0x11) / 255.0
        let b = CGFloat((v & 0xf) * 0x11) / 255.0
        return NSColor(srgbRed: r, green: g, blue: b, alpha: 1)
    default:
        return nil
    }
}
